src/cargo/core/workspace.rs

```diff
@@ -1522,27 +1519,83 @@ impl<'gctx> Workspace<'gctx> {
             })
             .sorted()
             .take(5)
-            .collect();
+            .collect()
+    }
 
+    fn report_unknown_features_error(
+        &self,
+        specs: &[PackageIdSpec],
+        cli_features: &CliFeatures,
+        found_features: &BTreeSet<FeatureValue>,
+    ) -> CargoResult<()> {
         let unknown: Vec<_> = cli_features
             .features
             .difference(found_features)
             .map(|feature| feature.to_string())
             .sorted()
             .collect();
 
-        if suggestions.is_empty() {
-            bail!(
-                "none of the selected packages contains these features: {}",
+        let (selected_members, unselected_members): (Vec<_>, Vec<_>) = self
+            .members()
+            .partition(|member| specs.iter().any(|spec| spec.matches(member.package_id())));
+
+        let missing_packages_with_the_features = unselected_members
+            .into_iter()
+            .filter(|member| {
+                unknown
+                    .iter()
+                    .any(|feature| member.summary().features().contains_key(&**feature))
+            })
+            .map(|m| m.name())
+            .collect_vec();
+
+        let these_features = if unknown.len() == 1 {
+            "this feature"
+        } else {
+            "these features"
+        };
+        let mut msg = if let [singular] = &selected_members[..] {
+            format!(
+                "the package '{}' does not contain {these_features}: {}",
+                singular.name(),
                 unknown.join(", ")
-            );
+            )
         } else {
-            bail!(
-                "none of the selected packages contains these features: {}, did you mean: {}?",
-                unknown.join(", "),
-                suggestions.join(", ")
+            let names = selected_members.iter().map(|m| m.name()).join(", ");
+            format!("none of the selected packages contains {these_features}: {}\nselected packages: {names}", unknown.join(", "))
+        };
+
+        use std::fmt::Write;
+        if !missing_packages_with_the_features.is_empty() {
+            write!(
+                &mut msg,
+                "\nhelp: include package{} with the missing feature{}:",
+                if missing_packages_with_the_features.len() != 1 {
+                    "s"
+                } else {
+                    ""
+                },
+                if unknown.len() != 1 { "s" } else { "" }
+            )?;
+            for member in &missing_packages_with_the_features {
+                write!(&mut msg, " --package {member}")?;
+            }
+        } else {
```

Comment:
Should we show both hints?

Reply:
No, I don't think so. If the workspace has a package with this exact feature name, then it's very unlikely to be a misspelling.

I've made it like this after adding the tests, and seeing how silly it looks when I use the correct real feature name, and it suggests an unrelated one.

Reply:
Ok.  I had thought there were real world cases where we might want both but I can't reconstruct them at this time.  We can always change it later.